flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/impl/StaticPartitionFileSplitEnumerator.java

```diff
@@ -0,0 +1,154 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.file.src.impl;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.connector.source.SourceEvent;
+import org.apache.flink.api.connector.source.SplitEnumerator;
+import org.apache.flink.api.connector.source.SplitEnumeratorContext;
+import org.apache.flink.connector.file.src.FileSource;
+import org.apache.flink.connector.file.src.FileSourceSplit;
+import org.apache.flink.connector.file.src.PendingSplitsCheckpoint;
+import org.apache.flink.connector.file.src.assigners.FileSplitAssigner;
+import org.apache.flink.connector.file.src.enumerate.FileEnumerator;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * An improved SplitEnumerator implementation for bounded / batch {@link FileSource} input. This
+ * enumerator uses the iterator of splits to allocate small batches of FileSourceSplits, which helps
+ * to avoid Java's out-of-memory (OOM) issue. All files present in the configured input directories
+ * are assigned to the readers. Once all files are processed, the source is finished. This class has
+ * a thin implementation. The actual logic for creating the set of FileSourceSplits
+ *
+ * <p>to process, and the logic to decide which reader gets what split, are still in {@link
+ * FileEnumerator} and in {@link FileSplitAssigner}, respectively.
+ */
+@Internal
+public class StaticPartitionFileSplitEnumerator
+        implements SplitEnumerator<FileSourceSplit, PendingSplitsCheckpoint<FileSourceSplit>> {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(StaticPartitionFileSplitEnumerator.class);
+
+    private final SplitEnumeratorContext<FileSourceSplit> context;
+
+    private final Iterator<List<FileSourceSplit>> splitsIterator;
+
+    private final FileSplitAssigner.Provider assignerFactory;
+
+    private FileSplitAssigner splitAssigner;
+
+    // ------------------------------------------------------------------------
+
+    public StaticPartitionFileSplitEnumerator(
```

Comment:
I'm wondering how the HiveSoure will use `StaticPartitionFileSplitEnumerator`?

Reply:
Do you mean invoke `createInputSplitsIterator` to create `StaticPartitionFileSplitEnumerator` in method `HiveSource#createEnumerator` or others? Do do intend to modify the interface `FileEnumerator#enumerateSplits`
to make it return `Iterator<List<FileSourceSplit>>`?

Also, for `StaticPartitionFileSplitEnumerator#snapshotState`,  seems it will only snapshot `remainingSplits` in `splitAssigner` and one next element of `splitsIterator`. How the source can restore from that? I mean isn't needed to know the partitions the `StaticPartitionFileSplitEnumerator` has processed so that it can skip these partition after restore?